test/FIFOMarket.test.ts

```diff
@@ -0,0 +1,47 @@
+import { expect, hardhat } from '@/test/helpers';
+
+const setupTest = hardhat.deployments.createFixture(async (hre) => {
+  const { getNamedAccounts, upgrades, ethers, run } = hre;
+  await run('deploy:erc1820');
+  const { noriWallet } = await getNamedAccounts();
+  const NORI = await ethers.getContractFactory('NORI');
+  const Removal = await ethers.getContractFactory('Removal');
+  const Certificate = await ethers.getContractFactory('Certificate');
+  const FIFOMarket = await ethers.getContractFactory('FIFOMarket');
+  const noriInstance = await upgrades.deployProxy(NORI, []);
+  const removalInstance = await upgrades.deployProxy(Removal, [], {
+    initializer: 'initialize()',
+  });
+  const certificateInstance = await upgrades.deployProxy(Certificate, [], {
+    initializer: 'initialize()',
+  });
+  const fifoMarketInstance = await upgrades.deployProxy(FIFOMarket, [
+    removalInstance.address,
+    noriInstance.address,
+    certificateInstance.address,
+    noriWallet,
+    15,
+  ]);
+  return {
+    contracts: {
+      NORI: noriInstance,
+      Removal: removalInstance,
+      Certificate: certificateInstance,
+      FIFOMarket: fifoMarketInstance,
+    },
+  };
+});
+
+describe('FIFOMarket', () => {
+  describe('Buying', () => {
+    it('should purchase removals and create a certificate', async () => {
+      const {
+        contracts: { NORI, Removal, Certificate, FIFOMarket },
+      } = await setupTest();
+      expect(NORI.address).to.be.properAddress;
+      expect(Removal.address).to.be.properAddress;
+      expect(Certificate.address).to.be.properAddress;
+      expect(FIFOMarket.address).to.be.properAddress;
+    });
+  });
```

Comment:
should we include a test about what happens if there aren't enough removals to fulfill the order?  or is it just too obvious that the transaction will revert and it's not interesting?

Reply:
we should definitely add tests for that!

For now I'm opting to handle those incrementally as we work through more requirements